### PR TITLE
update: generation from spec-nodeset, node.py, ua_binary.py

### PR DESCRIPTION
**Pull request by AndreasHeine**
according to https://github.com/FreeOpcUa/opcua-asyncio/pull/211

### CODE REVIEW

**Comment by oroulet** [comment]
tests fail. look at log looks like easy stuff

**Comment by AndreasHeine** [comment]
i know ... had no time today sorry! i am somete in the middle 😬 gibe me some time

**Comment by AndreasHeine** [comment]
messed somethin up... took a while to find but still not done!

**Comment by AndreasHeine** [comment]
@oroulet could you help me on that ?

```
    tree = ET.parse(path)
  File "/home/scrutinizer/.pyenv/versions/3.7.2/lib/python3.7/xml/etree/ElementTree.py", line 1197, in parse
    tree.parse(source, parser)
  File "/home/scrutinizer/.pyenv/versions/3.7.2/lib/python3.7/xml/etree/ElementTree.py", line 587, in parse
    source = open(source, "rb")
FileNotFoundError: [Errno 2] No such file or directory: '/home/scrutinizer/build/schemas/Opc.Ua.NodeSet2.xml'

----------------------------------------------------------------------
Ran 408 tests in 96.872s

FAILED (errors=1)
```
```

======================================================================

ERROR: test_std_address_space_references (tests_standard_address_space.StandardAddressSpaceTests)

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/home/travis/build/FreeOpcUa/python-opcua/tests/tests_standard_address_space.py", line 58, in test_std_address_space_references

    std_nodes = read_nodes(os.path.abspath(os.path.join(os.path.dirname(__file__), '..', 'schemas', 'Opc.Ua.NodeSet2.xml')))

  File "/home/travis/build/FreeOpcUa/python-opcua/tests/tests_standard_address_space.py", line 33, in read_nodes

    tree = ET.parse(path)

  File "/opt/python/2.7.15/lib/python2.7/xml/etree/ElementTree.py", line 1182, in parse

    tree.parse(source, parser)

  File "/opt/python/2.7.15/lib/python2.7/xml/etree/ElementTree.py", line 647, in parse

    source = open(source, "rb")

IOError: [Errno 2] No such file or directory: '/home/travis/build/FreeOpcUa/python-opcua/schemas/Opc.Ua.NodeSet2.xml'

----------------------------------------------------------------------

Ran 408 tests in 82.993s

FAILED (errors=1)

The command "./run-tests.sh" exited with 1.

Done. Your build exited with 1.
```

**Comment by AndreasHeine** [comment]
finally :)

**Comment by AndreasHeine** [comment]
@oroulet if you can afford some time to review my changes, go ahead!

File: opcua/ua/status_codes.py [hunk 1]
```diff
@@ -504,12 +504,13 @@ class StatusCodes(object):
...
+        if val in code_to_name_doc:
+            return code_to_name_doc[val]
         else:
-            return 'Good', 'Unknown StatusCode value: {}'.format(val)
+            if val & 1 << 31:
+                return 'Bad', 'Unknown StatusCode value: {}'.format(val)
+            elif val & 1 << 30:
+                return 'UncertainIn', 'Unknown StatusCode value: {}'.format(val)
+            else:
+                return 'Good', 'Unknown StatusCode value: {}'.format(val)
```
Comment (swamper123) [inline opcua/ua/status_codes.py:515]: Maybe these could be converted to f-strings, to get format-free step by step.

File: opcua/ua/ua_binary.py [hunk 2]
```diff
@@ -140,6 +140,9 @@ def unpack(self, data):
     def pack_array(self, data):
         if data is None:
             return Primitives.Int32.pack(-1)
+        if not isinstance(data, list):
+            logger.warning('ua_binary.py > _Primitive1 > pack_array > data: {0} is not a instance of "list"!'.format(data))
```
Comment (swamper123) [inline opcua/ua/ua_binary.py:144]: The same here with f-strings

File: schemas/generate_address_space.py [hunk 3]
```diff
@@ -3,34 +3,37 @@
...
 """
 import sys
+import os
+import datetime
 import logging
 # sys.path.insert(0, "..")  # load local freeopcua implementation
 #from opcua import xmlparser
 import opcua.common.xmlparser as xmlparser
 
+BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
```
Comment (swamper123) [inline schemas/generate_address_space.py:13]: os is a "heavy" library, we may should use pathlib for such cases.

File: schemas/generate_address_space.py [hunk 4]
```diff
@@ -323,22 +330,23 @@ def make_refs_code(self, obj, indent):
 
 
 def save_aspace_to_disk():
-    import os.path
-    path = os.path.join("..", "opcua", "binary_address_space.pickle")
-    print("Savind standard address space to:", path)
-    sys.path.append("..")
+    path = os.path.join(BASE_DIR, 'opcua', 'binary_address_space.pickle')
```
Comment (swamper123) [inline schemas/generate_address_space.py:333]: The same here with pathlib if possible.

File: schemas/generate_event_objects.py [hunk 5]
```diff
@@ -23,18 +27,25 @@ def write(self, line):
...
+        self.write("")
+        names = ", ".join(f'"{event.browseName}"' for event in events)
+        self.write(f"__all__ = [{names}]")
+        self.write("")
 
     def addProperties(self, event):
         for ref in event.references:
             if ref.referenceType == "HasProperty":
-                    self.write("self.add_property('{0}', {1}, {2})".format(ref.refBrowseName, self.getPropertyValue(ref), self.getPropertyDataType(ref)))
+                self.write("self.add_property('{0}', {1}, {2})".format(
```
Comment (swamper123) [inline schemas/generate_event_objects.py:45]: Format -> fstrings

File: schemas/generate_model_event.py [hunk 6]
```diff
@@ -79,8 +84,12 @@ def findNodeWithNodeId(self,root, nodeId):
         return node
 
     def checkNodeType(self, node):
-        if (node.tag == self.nameSpace + "UAObjectType") or (node.tag == self.nameSpace + "UAVariable") or (
-            node.tag == self.nameSpace + "UAObject") or (node.tag == self.nameSpace + "UAMethod") or (node.tag == self.nameSpace + "UAVariableType"):
+        if (
+                (node.tag == self.nameSpace + "UAObjectType") or
```
Comment (swamper123) [inline schemas/generate_model_event.py:88]: I am not sure what would be faster/better, a compare "list" like this or using something with self.nameSpace.endswith(("UAObjectType","UAVariable", "UAObject","UAMethod","UAVariableType")).
It feels not DRY.

File: opcua/ua/ua_binary.py (hunk 2 above)
Comment (AndreasHeine) [reply opcua/ua/ua_binary.py:144]: No! to be compatible to python 2.7 ...

File: schemas/generate_address_space.py (hunk 3 above)
Comment (AndreasHeine) [reply schemas/generate_address_space.py:13]: @oroulet mentiond it in asyncua but i was more used to os.path but yes we can!

File: schemas/generate_event_objects.py (hunk 5 above)
Comment (AndreasHeine) [reply schemas/generate_event_objects.py:45]: Yes this one could be replaced!

File: opcua/common/event_objects.py [hunk 7]
```diff
@@ -1,10 +1,13 @@
...
...ngeEvent", "GeneralModelChangeEvent", "TransitionEvent", "AuditUpdateStateEvent", "ProgramTransitionEvent", "SemanticChangeEvent", "AuditUrlMismatchEvent", "RefreshStartEvent", "RefreshEndEvent", "RefreshRequiredEvent", "AuditConditionEvent", "AuditConditionEnableEvent", "AuditConditionCommentEvent", "AuditHistoryEventUpdateEvent", "AuditHistoryValueUpdateEvent", "AuditHistoryDeleteEvent", "AuditHistoryRawModifyDeleteEvent", "AuditHistoryAtTimeDeleteEvent", "AuditHistoryEventDeleteEvent", "EventQueueOverflowEvent", "ProgramTransitionAuditEvent", "AuditConditionRespondEvent", "AuditConditionAcknowledgeEvent", "AuditConditionConfirmEvent", "AuditConditionShelvingEvent", "ProgressEvent", "SystemStatusChangeEvent", "AuditProgramTransitionEvent", "TrustListUpdatedAuditEvent", "CertificateUpdatedAuditEvent", "AuditConditionResetEvent", "PubSubStatusEvent", "PubSubTransportLimitsExceedEvent", "PubSubCommunicationFailureEvent", "AuditConditionSuppressionEvent", "AuditConditionSilenceEvent", "AuditConditionOutOfServiceEvent", "RoleMappingRuleChangedAuditEvent", "KeyCredentialAuditEvent", "KeyCredentialUpdatedAuditEvent", "KeyCredentialDeletedAuditEvent", "AuditHistoryAnnotationUpdateEvent"]
```
Comment (oroulet) [inline opcua/common/event_objects.py:8]: why do we need that one? do we want to hide something?

File: schemas/generate_address_space.py (hunk 3 above)
Comment (oroulet) [reply schemas/generate_address_space.py:13]: this is not part of core library so not so important

File: schemas/generate_address_space.py [hunk 8]
```diff
@@ -3,34 +3,37 @@
...
+        return f"NodeId.from_string('{val}')"
     return ua_type_to_python(val, _get_uatype_name(cls, attr))
 
 
 def _get_uatype_name(cls, attname):
     for name, uat in cls.ua_types:
         if name == attname:
             return uat
-    raise Exception("Could not find attribute {} in obj {}".format(attname, cls))
+    raise Exception(f"Could not find attribute {attname} in obj {cls}")
```
Comment (oroulet) [inline schemas/generate_address_space.py:29]: OK that kind of changes make that file unsuable with python 2.7 and earlier 3 version. This is fine for generate_xxx files but not for the library code

File: opcua/common/event_objects.py (hunk 7 above)
Comment (AndreasHeine) [reply opcua/common/event_objects.py:8]: to be honest i compared asyncua and python-opcua side by side and try to make them as equal as possible... i mean it should be equal (except opcua/asyncua)

File: schemas/generate_model_event.py (hunk 6 above)
Comment (oroulet) [reply schemas/generate_model_event.py:88]: this is not part of lirbary so need to need to opitmize. This is just run by library developers

**Comment by AndreasHeine** [comment]
decide f-string yes or no! then we can make it 2.7 compliant it would take me some time... i could do it but in this turn asyncua should also be edited in this case to f-strings only !?

**Comment by AndreasHeine** [comment]
as far is i know the reason for asyncua was python >3.6 asyncio module guess some dev's still use or maintain products with 2.7!?
but on the other hand its just the generation from spec. files which does not work on 2.7 atm.

**Comment by oroulet** [comment]
f-string for generate-xx file is fine.

**Comment by oroulet** [comment]
and of course f-string is best for asyncua

**Comment by AndreasHeine** [comment]
ok i try to recap:
asyncua -> f-strings everywhere -> will rework the files with new PR
python-opcua -> f-strings only for generation? to make it consistent in both?

**Comment by swamper123** [comment]
@AndreasHeine I would say, if devs are still counting on python2, since the EOL was announced years ago, they are on their own. Maybe adding a tag (latest-python2-support or sth. similar) should be added and we step forward in version support.

**Comment by AndreasHeine** [comment]
@swamper123 : Its less work to remove f-strings from generation files (python-opcua only) then changing all in python-opcua to f-string!? i would feel mad if we would reduse compatibility just for f-strings!?

i will rework asyncua generation to f-string only!
what happens to python-opcua !? @oroulet still has not decided he can merge if he want and we make a cleanup PR afterwards...

the pathlib thing is also possible, my tought was just its for generation only and not part of the opcua related runtime code so i actually dont care about the module size that much.

for me its ok anyways! its just the generation of the low level stuff so as far as we have the latest generated, it should be fine.

now i leave you up for a decision!

**Comment by swamper123** [comment]
> @swamper123 : Its less work to remove f-strings from generation files (python-opcua only) then changing all in python-opcua to f-string!? i would feel mad if we would reduse compatibility just for f-strings!?

Hmm you may be right with that. I am just thinking about other future problems as well, but maybe keeping python2 until a bigger break/major performance boost in the python3.x/4.x versions would be the better idea.

But in the end @oroulet has to decide as well.

**Comment by oroulet** [comment]
The original goal is to do as litle as possible with python-opcua. asyncua has a sync wrapper that does most of it, it just needs more testing to find out f there are bugs or performance issues.

so yes

asyncua -> f-strings everywhere -
python-opcua -> f-strings only for generation but do not spend too much time on that

**Comment by AndreasHeine** [comment]
then merge and we focus on asyncua!

**Pull request merged**